pkg/operator/controller/status.go

```diff
@@ -95,123 +94,147 @@ func (r *reconciler) syncOperatorStatus() error {
+func (r *reconciler) getOperatorState() (*corev1.Namespace, []operatorv1.DNS, error) {
 	ns := manifests.DNSNamespace()
 	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
 		if errors.IsNotFound(err) {
-			return nil, nil, nil, nil
+			return nil, nil, nil
 		}
-		return nil, nil, nil, fmt.Errorf("failed to get namespace %s: %v", ns.Name, err)
+		return nil, nil, fmt.Errorf("failed to get namespace %s: %v", ns.Name, err)
 	}
 
 	dnsList := &operatorv1.DNSList{}
 	if err := r.client.List(context.TODO(), dnsList); err != nil {
-		return nil, nil, nil, fmt.Errorf("failed to list dnses: %v", err)
+		return nil, nil, fmt.Errorf("failed to list dnses: %v", err)
 	}
 
-	daemonsetList := &appsv1.DaemonSetList{}
-	if err := r.client.List(context.TODO(), daemonsetList, client.InNamespace(ns.Name)); err != nil {
-		return nil, nil, nil, fmt.Errorf("failed to list daemonsets: %v", err)
+	return ns, dnsList.Items, nil
+}
+
+// computeOperatorStatusConditions computes the operator's current state.
+func computeOperatorStatusConditions(oldConditions []configv1.ClusterOperatorStatusCondition, ns *corev1.Namespace,
+	dnsesAvailable, dnses int) []configv1.ClusterOperatorStatusCondition {
+	var oldDegradedCondition, oldProgressingCondition, oldAvailableCondition *configv1.ClusterOperatorStatusCondition
+	for i := range oldConditions {
+		switch oldConditions[i].Type {
+		case configv1.OperatorDegraded:
+			oldDegradedCondition = &oldConditions[i]
+		case configv1.OperatorProgressing:
+			oldProgressingCondition = &oldConditions[i]
+		case configv1.OperatorAvailable:
+			oldAvailableCondition = &oldConditions[i]
+		}
 	}
-	return ns, dnsList.Items, daemonsetList.Items, nil
+
+	conditions := []configv1.ClusterOperatorStatusCondition{}
+	conditions = append(conditions, computeOperatorDegradedCondition(oldDegradedCondition, dnsesAvailable, dnses, ns))
+	conditions = append(conditions, computeOperatorProgressingCondition(oldProgressingCondition, dnsesAvailable, dnses))
+	conditions = append(conditions, computeOperatorAvailableCondition(oldAvailableCondition, dnsesAvailable))
+
+	return conditions
 }
 
-// computeStatusConditions computes the operator's current state.
-func computeStatusConditions(conditions []configv1.ClusterOperatorStatusCondition, ns *corev1.Namespace, dnses []operatorv1.DNS, daemonsets []appsv1.DaemonSet) []configv1.ClusterOperatorStatusCondition {
+// computeOperatorDegradedCondition computes the operator's current Degraded status state.
+func computeOperatorDegradedCondition(oldCondition *configv1.ClusterOperatorStatusCondition, dnsesAvailable, dnsesTotal int,
+	ns *corev1.Namespace) configv1.ClusterOperatorStatusCondition {
 	degradedCondition := &configv1.ClusterOperatorStatusCondition{
-		Type:   configv1.OperatorDegraded,
-		Status: configv1.ConditionUnknown,
+		Type: configv1.OperatorDegraded,
 	}
-	if ns == nil {
+	switch {
+	case dnsesAvailable == 0 || dnsesAvailable != dnsesTotal && ns == nil:
```

Comment:
`&&` is higher precedence than `||`, so this expression groups as follows: `dnsesAvailable == 0 || (dnsesAvailable != dnsesTotal && ns == nil)`.

Reply:
Since the expression starts to get hairy with the three subexpressions, I'd suggest simplifying by putting `case ns == nil:` first so that subsequent cases can assume the `ns` is not nil.  If the namespace does not exist, then DNSes cannot exist, so I believe it is sufficient just to say that the namespace does not exist if that is the case.

Reply:
Good catch